### ISSUE TITLE
Improve re-downloading files when the previous download was stopped mid-way

### ISSUE THREAD

**Issue by kan-fu** [OceanNetworksCanada/api-python-client#54]
When downloading files was interrupted for some unknown reasons (internet connection issue or server response issue), the file is kept in the outPath directory, with the file size smaller than the expected size. When the user wants to resume the download, they have no idea if the file is downloaded successfully or not. This becomes a pain if the script downloads a lot of files, and some of the files are corrupted.

Check the email for the detailed description:

> I’ve been running the archive download script to download large quantities of audio files from MEF. This has been working well, but at the request of our IT admin I have placed the downloader on a schedule using cron on linux such that the downloading only occurs during off-peak hours (so I don’t hog all the bandwidth… apparently that could be an issue).
> 
> The issue with this is that the process sometimes terminates mid-way through an audio file being written, and I end up with some files with 0 size or smaller than they should be (see screenshot below). When I resume the download, the onc library seems to just scan for the existing file names in the directory in order to determine which files have already been downloaded. It doesn’t appear to look at the size of the files compared with their size on ONC’s server to determine if the files were downloaded correctly/completely.
> 
>  
> 
> Is this something that could be added to the onc python library? So that rather than just checking file names to see what has been downloaded, it also checked file size or duration compared with what it is supposed to be, to ensure the files have been downloaded in their entirety? My current workaround plan is to write a script which goes through all the files later, checks for those with a suspiciously small file size, and re-downloads them from onc. But it would be great to avoid that if possible!

**Comment by kan-fu** [comment]
There are two ways to approach the issue:
1. Compare the file size from the server with the local version if the file exists, which is suggested in the email.  
2. Make the invalid files easier to identify.

The first solution is not easy. `/archivefile/download` returns `text/plain` response, and the response header does not contain any information about the content size. `/archivefile` discovery service does return file size, but it needs device information (like https://data.oceannetworks.ca/apiproxy/archivefile/device?deviceCode=UA3DCAMERA&returnOptions=all), so it is hard to call it with only file name. Also calling an extra web service for file size just seems not right.
```
# Response header 
{
    "Server": "nginx/1.14.1",
    "Date": "Mon, 09 Dec 2024 18:16:03 GMT",
    "Content-Type": "text/plain;charset=ISO-8859-1",
    "Connection": "keep-alive",
    "Vary": "Accept-Encoding",
    "Set-Cookie": "JSESSIONID=B3C35E7A932BFA7F11E5166D6CFA3C77; Path=/; HttpOnly",
    "Access-Control-Allow-Headers": "X-Requested-With, Content-Type",
    "Access-Control-Allow-Credentials": "true",
    "Last-Modified": "Mon, 09 Dec 2024 18:16:03 GMT",
    "Cache-Control": "public",
    "Pragma": "no-cache",
    "Content-Disposition": "attachement; filename=BPR-Folger-59_20191123T000000.000Z.txt",
    "Strict-Transport-Security": "max-age=31536000; includeSubDomains",
    "Content-Encoding": "gzip",
}
```
In #43 we added the `stream=True` to skip the download if the file exists. It defeats the purpose to download it again just for the file size. So basically this solution requires  `/archivefile/download` to return file size in the header.

The second option seems better for me. The issue is tricky because the invalid files have a smaller file size, which is hard to tell without knowing the actual file size. Things will be much easier if all the invalid files have 0 file size (At least users can manually delete those 0-size files first). In #43 I replaced response.content with [`response.iter_content`](https://github.com/OceanNetworksCanada/api-python-client/pull/43/files#diff-06c4337f627f27285b84f226d7be19fd43ea73168efb95f1ed17f9b46d798200R27), as suggested at [here](https://requests.readthedocs.io/en/latest/user/quickstart/#raw-response-content). I guess by that I introduced the bug to make the file size smaller when things go wrong. Before that PR the download is in a bulk to memory first, so it should be either full size or 0 byte size. I can revert it back, and also add the logic [here](https://github.com/OceanNetworksCanada/api-python-client/blob/main/src/onc/modules/_util.py#L21) so that it is 
`if Path.exists(filePath) and os.path.getsize(filePath) != 0 and not overwrite:` when throwing `FileExistsError`.
I am not 100% sure if there are other cases to make the invalid files have a non-zero file size (shouldn't if I understand it correctly). But reverting it back is a good start.
@aschlesin Could you send an email to the user to see if the comment makes sense to them? They can reply here if they have comments or suggestions.

**Comment by bsmithacoustics** [comment]
@kan-fu I agree that the second option would be sufficient. I'd be fine with manually deleting the 0 size files. Just a thought - could the library check for 0 size files and try to re-download them automatically? If not I am fine with your suggestion of deleting the 0 size files and rerunning my script, but it would be convenient to avoid the manual step if possible. Thank you.

**Comment by kan-fu** [comment]
> @kan-fu I agree that the second option would be sufficient. I'd be fine with manually deleting the 0 size files. Just a thought - could the library check for 0 size files and try to re-download them automatically? If not I am fine with your suggestion of deleting the 0 size files and rerunning my script, but it would be convenient to avoid the manual step if possible. Thank you.

I think that is what `if Path.exists(filePath) and os.path.getsize(filePath) != 0 and not overwrite:` does when throwing the `FileExistsError` exception (right now it has no `os.path.getsize(filePath) != 0`), meaning if the file has 0 file size, it will not throw the exception. So you don't need to delete those 0-size files before rerunning the script.

For example, We have file XXX.jpg that has 0 file size, and you call `downloadArchivefile("XXX.jpg", overwrite=False)`. The if statement will not be executed, so the file will still be downloaded and overwritten. This is based on the scenario that the filename (or filter if `downloadDirectArchivefile` is used) is known . I am thinking about the use case that the user used some script to download some files, found some files having 0 file size, and rerun the script. Files that do not have 0 file size will be skipped, and everything else will be downloaded again.

If the filename is not known, it is also easy to write a small script to get all the files that have 0 file size, and call `downloadArchivefile(filename)`. Might be a rare case, but it is doable.

Does it answer your question? @bsmithacoustics

**Comment by bsmithacoustics** [comment]
@kan-fu Yes - that all sounds good to me!

**Comment by MHee** [comment]
I  think the most efficient and reliable way to plainly download a large number of archived files from Oceans 3.0 is to use an existing download manager that allows for accepting and managing jobs, programmatically.

Maybe, we can add some guide on how to use one.

It seems that [aria2](https://aria2.github.io/manual/en/html/README.html) ticks all the boxes. It is open source and a single executable that is available for most operating systems. 

You can launch the executable file at the root of your desired data directory and use the following options for the program to wait and listen for download jobs:

`aria2c --enable-rpc=true --rpc-listen-all=false --save-session=session.txt --save-session-interval=900 --rpc-secret=aria2_token --log-level=notice --log=downloads.log --console-log-level=warn`

This is the part of the documentation that describes how to interact with it  
[https://aria2.github.io/manual/en/html/aria2c.html\#rpc-options](https://aria2.github.io/manual/en/html/aria2c.html#rpc-options)

And the following  gist:

[https://gist.github.com/MHee/dadbabd9e4d496b74092879ae43d6061](https://gist.github.com/MHee/dadbabd9e4d496b74092879ae43d6061)

Shows how to

* Find what locations around Endeavour have currently hydrophones installed  
* Get a list of recent hydrophone files  
* Send download requests to the running aria2 instance  
* Interact with the instance

In my tests, I used the default of five workers and was able to saturate my Wi-Fi card/Network at home pretty good:  

I was able to download 24 hrs worth of all files for a single hydrophone in about 3 minutes:  

If that 2GB/minute speed is representative, it would take about 8 hrs per TB.  
That seems to be pretty good. We should just make sure that you are not overwhelming our servers and get blocked. 

### Useful parameters

**\-j, \--max-concurrent-downloads=\<N\>** : Number of files downloaded at the same time (default 5\)

**\--max-download-result=\<NUM\>**

**\-l, \--log=\<LOG\>**[¶](https://aria2.github.io/manual/en/html/aria2c.html#cmdoption-l)

**\--log-level=\<LEVEL\>**

Not Relevant for download from single server  
**\-s, \--split=\<N\>** :  Connections per download  
**\-x, \--max-connection-per-server=\<NUM\>**  : Per download

**Comment by MHee** [comment]
Aria2 benefits:
* Fast
* Easy on resources
* Optional GUI to manage downloads later (if required)
* Can limit download speed if wanted
* Can reprioritize downloads
* Keeps track of finished files and can continue partial downloads
* Can write a list of pending jobs when terminated, can start with the list when re-started

**Comment by MHee** [comment]
Many download managers append a temporary suffix to the file while it is downloaded and rename the file to its final name only after the download has completed successfully.

**Comment by kan-fu** [comment]
Based on their [website](https://aria2.github.io/), aria2c is very easy to use. For example I can run `aria2c "https://data.oceannetworks.ca/api/archivefile/download?filename=BPR-Folger-59_20191123T000000.000Z.txt&token=TOKEN"` in the terminal to download one file. 

Using download manager like aria2 is a good idea, especially downloading big-size files, but it is outside of the scope of this client library.

To make onc client library work with aria2 for archivefiles (not very easy to integrate with data product delivery), a new feature can be added to just return the download link given the filename (`getArchivefile`) or filter(`downloadDirectArchivefile`). Users can easily run `aria2c link` or `aria2c -Z link1 link2` if they wish to use aria2.

**Comment by MHee** [comment]
@kan-fu, I agree that the implementation is out of the scope of the client library, but we could provide an example, or a small in depended repository (or clean up my gist). I think the real kicker of aria2c is the RPC interface.

You can just run it as server on the computer with the big disk and send jobs to it: 
![image](https://github.com/user-attachments/assets/c93fecd1-5235-4851-b182-d45a1bf8d63d)

**Comment by kan-fu** [comment]
There is a webui for aria2c at https://github.com/ziahamza/webui-aria2 which is based on the rpc server but is  more user friendly. It looks more like a standard download manager like IDM.

I think it is a good idea to introduce this idea in our documentation website. I will implement the archivefile link-generation (basically `generate_download_url` in your gist) feature and add examples to explain how I would do that.

**Comment by kan-fu** [comment]
Hi @bsmithacoustics, it is near the Christmas break, so the PR might not be merged until early Jan. If you want to try the new download, you can pip install from the git branch by using `pip install git+https://github.com/OceanNetworksCanada/api-python-client.git@issue-54-revert-download`.
